File: mns-reactor/src/main/java/cz/fb/manaus/reactor/categorizer/CoverageCategorizer.java
package cz.fb.manaus.reactor.categorizer;


import com.google.common.base.CaseFormat;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;
import cz.fb.manaus.core.category.BetCoverage;
import cz.fb.manaus.core.category.categorizer.SettledBetCategorizer;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.SettledBet;
import cz.fb.manaus.core.model.Side;
import org.springframework.stereotype.Component;

import java.util.EnumSet;
import java.util.Map;
import java.util.Set;

@Component
public class CoverageCategorizer implements SettledBetCategorizer {

    public static final String PREFIX = "coverage_";

    @Override
    public boolean isSimulationSupported() {
        return false;
    }

    @Override
    public Set<String> getCategories(SettledBet settledBet, BetCoverage coverage) {
        var marketId = settledBet.getBetAction().getMarket().getId();
        var selectionId = settledBet.getSelectionId();
        var sides = coverage.getSides(marketId, selectionId);
        Preconditions.checkState(sides.size() > 0);
        var builder = ImmutableMap.<Side, Double>builder();
        for (var side : sides) {
            builder.put(side, coverage.getAmount(marketId, selectionId, side));
        }
        var amounts = builder.build();
        return getCategories(settledBet.getPrice().getSide(), amounts);
    }

    Set<String> getCategories(Side mySide, Map<Side, Double> amounts) {
        var sideFormatted = CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.UPPER_CAMEL, mySide.name());
        if (EnumSet.of(mySide).equals(amounts.keySet())) {
            var soloSide = "solo" + sideFormatted;
            return Set.of(PREFIX + soloSide, PREFIX + "solo");
        } else if (EnumSet.of(mySide, mySide.getOpposite()).equals(amounts.keySet())) {
            var builder = ImmutableSet.<String>builder().add(PREFIX + "both");
            builder.add(PREFIX + "both");
            builder.add(PREFIX + "both" + sideFormatted);
            if (Price.amountEq(amounts.get(Side.LAY), amounts.get(Side.BACK))) {
                builder.add(PREFIX + "bothEqual");
            } else if (amounts.get(Side.LAY) > amounts.get(Side.BACK)) {
                builder.add(PREFIX + "bothLayGt");
            } else if (amounts.get(Side.LAY) < amounts.get(Side.BACK)) {
                builder.add(PREFIX + "bothBackGt");
            }
            return builder.build();
        }
        throw new IllegalStateException();
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/categorizer/CoverageCategorizer.kt
package cz.fb.manaus.reactor.categorizer


import com.google.common.base.CaseFormat
import cz.fb.manaus.core.category.BetCoverage
import cz.fb.manaus.core.category.categorizer.RealizedBetCategorizer
import cz.fb.manaus.core.model.RealizedBet
import cz.fb.manaus.core.model.Side
import cz.fb.manaus.core.model.amountEq
import org.springframework.stereotype.Component
import java.util.*

@Component
class CoverageCategorizer : RealizedBetCategorizer {

    override val isSimulationSupported: Boolean = false

    override fun getCategories(realizedBet: RealizedBet, coverage: BetCoverage): Set<String> {
        val marketId = realizedBet.market.id
        val selectionId = realizedBet.settledBet.selectionId
        val sides = coverage.getSides(marketId, selectionId)
        check(sides.isNotEmpty())
        val amounts = sides.map { it to coverage.getAmount(marketId, selectionId, it) }.toMap()
        return getCategories(realizedBet.settledBet.price.side, amounts)
    }

    internal fun getCategories(mySide: Side, amounts: Map<Side, Double>): Set<String> {
        val sideFormatted = CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.UPPER_CAMEL, mySide.name)
        if (EnumSet.of(mySide) == amounts.keys) {
            val soloSide = "solo$sideFormatted"
            return setOf(PREFIX + soloSide, PREFIX + "solo")
        } else if (EnumSet.of(mySide, mySide.opposite) == amounts.keys) {
            val builder = mutableSetOf<String>()
            builder.add(PREFIX + "both")
            builder.add(PREFIX + "both")
            builder.add(PREFIX + "both" + sideFormatted)
            when {
                amounts[Side.LAY]!! amountEq amounts[Side.BACK]!! -> builder.add(PREFIX + "bothEqual")
                amounts[Side.LAY]!! > amounts[Side.BACK]!! -> builder.add(PREFIX + "bothLayGt")
                amounts[Side.LAY]!! < amounts[Side.BACK]!! -> builder.add(PREFIX + "bothBackGt")
            }
            return builder.toSet()
        }
        throw IllegalStateException()
    }

    companion object {
        const val PREFIX = "coverage_"
    }
}
